teps/0015-pause-pipeline.md

```diff
@@ -0,0 +1,122 @@
+transparent to the `PipelineRun` controller and `PipelineRun` creator (e.g.,
+Tekton triggers).
+
+With this feature, platforms can control when `PipelineRuns` are started without
+introducing a bespoke system for storing `PipelineRuns` - potentially adding new
+failure scenarios or UI complexity.
+
+### Goals
+
+* Provide a pause mechanism for `PipelineRuns` that operators can use
+  to control whether or not a `PipelineRun` is running (e.g., "creating new
+  `TaskRuns`").
+
+### Non-Goals
+
+* Support pausing of a running `TaskRun` as that would requiring pausing the
+  Kubernetes pods themselves.
+
+## Requirements
+
+* Provide a mechanism to operators that can be used to pause a `PipelineRun`.
+* `PipelineRuns` that are created with the pause setting enabled are not started
+  until the setting is toggled off.
+* `PipelineRuns` that have the pause setting enabled after they are started will
+  not create any new `TaskRuns` until the setting is toggled off.
+
+## Proposal
+
+The `PipelineRunSpec` supports a field called `Status` of the type
+`PipelineRunSpecStatus`. Currently the field only supports a single setting,
+`PipelineRunCancelled`. This field will be expanded to support an additional
+setting, `PipelineRunPaused`.
+
+When the `Status` field is set to `PipelineRunPaused`, the Tekton reconciler
+will set the `Succeeded` `Condition` `Reason` in the `PipelineRun` `Status` to
+`PipelineRunPaused`:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "2020-09-15T18:15:29Z"
+    message: 'PipelineRun was paused.'
+    reason: PipelineRunPaused
+    status: "Unknown"
+    type: Succeeded
+```
+
+To "un-pause" a resource, the operator should set the `Status` field to empty,
+the Tekton controller will see the change and proceed with the Pipeline as
+normal. Upon unpausing, the `Succeeded` `Condition` `Reason` will be changed to
+`Running`:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "2020-09-15T19:46:34Z"
+    message: Not all Steps in the Task have finished executing
+    reason: Running
+    status: Unknown
+    type: Succeeded
+```
+
+### User Stories
+
+#### Platform Implementor
+
+As a platform implementor, I want to be able to control when a `PipelineRun` is
+able to run without introducing bespoke machinery.
+
+## Alternatives
+
+* Instead of adding this setting directly to the `PipelineRun` object,
+  users could introduce a new CRD, e.g., a `QueuedPipelineRun`, which has all
+  the same fields as a `PipelineRun` but is inert. This has a number of
+  drawbacks in that users would need to be aware of it, the UI would need to be
+  aware of it, and the added complexity of maintaining a new CRD or abstraction.
+* `PipelineRuns` that are not ready to be started could be stored in a
+  non-Kubernetes-native queue, such as Redis, and only submitted once they are
+  ready to run. This adds complexity in that a new data store must be run and
+  it is not easy to provide users visibility into the queue.
```

Comment:
another alternative we could add here is  instead of adding a field to `spec.status` we add a new condition type, e.g. in addition to condition "succeeded" we have a condition "pending" or something

i think that the current proposal is better because I think it reflects the purpose of spec vs. status better:

- spec = the desired behavior (e.g. don't actually run it)
- status = what is actually happening (e.g. it is not running)

Reply:
I agree withe the status `Pending`.

Actually the way I achieve `Paused` is add `PipelinePaused` to `spec.status` just like the way we achieve `PipelineCancelled`. Ref: tektoncd/pipeline#3223

Maybe I should write another TEP and change the name `Pause` to `Pending` in my PR.

Reply:
@FogDong tho they are similar I see these as 2 fairly different sets of functionality:

1. Pausing Runs = no matter what state the Run currently is in, stop it and don't allow it to progress further (which I think is what tektoncd/pipeline#3223 is about?)
2. Introduce a way to control initial scheduling of runs = Introduce a way to "freeze" Runs before they execute and make it possible for an external system to control when Runs are actually executed (which is what this "pending" TEP is about)

I would see having two different "levers" (e.g. 2 different values in spec.status) to control this. Or even if we use the same spec.status value for both, I think we'd likely approve and move forward with the proposal here, which has a much smaller scope.

To move forward with the pausing functionality we'll probably need to discuss in a further TEP - and especially to understand the use cases around this. (note that a "pause" of sorts could be introduced via a custom task [tep 0002](https://github.com/tektoncd/community/blob/master/teps/0002-custom-tasks.md)